File: asynchronous/src/main/java/com/dao/asynchronous/AsynchronousProviders.java
package com.dao.asynchronous;

import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

/**
 * Created in 09/11/18 17:24.
 *
 * @author Diogo Oliveira.
 */
public class AsynchronousProviders
{
    public static <T extends Asynchronous> T of(Class<T> asyncClass, FragmentManager manager)
    {
        T asynchronous = null;

        try
        {
            asynchronous = asyncClass.newInstance();
            Asynchronous running = (Asynchronous)manager.findFragmentByTag(asynchronous.tag());

            if((running == null) || running.isRemoving())
            {
                FragmentTransaction transaction = manager.beginTransaction();
                transaction.add(asynchronous, asynchronous.tag()).commitNowAllowingStateLoss();
            }
            else
            {
                //noinspection unchecked
                asynchronous = (T) running;
            }
        }
        catch(IllegalAccessException e)
        {
            e.printStackTrace();
        }
        catch(InstantiationException e)
        {
            e.printStackTrace();
        }

        return asynchronous;
    }
}

File: asynchronous/src/main/java/com/dao/asynchronous/AsynchronousProviders.kt
@file:Suppress("UNCHECKED_CAST")

package com.dao.asynchronous

import androidx.fragment.app.FragmentManager

/**
 * Created in 30/01/19 20:39.
 *
 * @author Diogo Oliveira.
 */
object AsynchronousProviders
{
    fun <T : Asynchronous<*, *>> of(asyncClass: Class<T>, manager: FragmentManager): T
    {
        var asynchronous: T? = null

        try
        {
            asynchronous = asyncClass.newInstance()
            val running = manager.findFragmentByTag(asynchronous!!.tag()) as Asynchronous<*, *>?

            if(running == null)
            {
                val transaction = manager.beginTransaction()
                transaction.add(asynchronous.async, asynchronous.tag()).commit()
            }
            else
            {

                asynchronous = running as T
            }
        }
        catch(e: IllegalAccessException)
        {
            e.printStackTrace()
        }
        catch(e: InstantiationException)
        {
            e.printStackTrace()
        }

        return asynchronous!!
    }
}